### ISSUE TITLE
Fitsignal output: Correct scaling and additional output

### ISSUE THREAD

**Issue by laenan8466** [JeschkeLab/DeerLab#86]
# Overview
This issue combines several reports concerning the fitsignal output (scaling) that can be tracked by this issue:

- [x] Output doesn't match input. Scale factor output (`fit.scale`) does not scale the input (`Vexp`) correctly to match the `fit.V` ouput.
    As described in https://github.com/JeschkeLab/DeerLab/issues/74#issuecomment-791638127
- [ ] User should have the choice between a scaled and not scaled output.
    This build up from https://github.com/JeschkeLab/DeerLab/pull/78#issuecomment-791894520. First attempt was made in [fix/fitsignal_plot](https://github.com/JeschkeLab/DeerLab/tree/fix/fitsignal_plot).
- [x]  Background CIs cannot be scaled from the given output in `fit.`
     This was mentioned in https://github.com/JeschkeLab/DeerLab/issues/74#issuecomment-791638127 as well.

As all issues are closely connected, I put them in this issue together. I propose to add all work in #78. 
The following part describes 1 and 3 in detail.

# Detailed descriptions

## Minimal Working example
I compared the output of `fit.plot()` and my 'manual' plot, using the same layout as in `fit.plot()`. This example will be used to describe both problems.

### Requirements
- For the tests I used https://github.com/JeschkeLab/DeerLab/commit/e36b6ba0f655379b5541a6226fc50a0616c2a417.
  - A small change is applied to `fitsignal.py`: *Remove/Comment line 611*:
    ```python
            # plt.show()
    ```
    This is important so that `axs` can be manipulated afterwards. I would recommend to skip the `plt.show()` command anyway, or manipulation of the returned axis isn't possible (at least for me). In my IDE (Spyder) plots are shown after running a script/command by default.
- Use the [two files in this MWE](https://github.com/JeschkeLab/DeerLab/files/6092939/LaserIMD_MWE.zip). They include a `LaserIMD_MWE.pkl` file with the data and the `LaserIMD_MWE.py` script.
- Run the `LaserIMD_MWE.py` script to get two Graphs.

#### Output from `fit.plot()`
![grafik](https://user-images.githubusercontent.com/21331242/110164906-582f8900-7df2-11eb-95f0-2a943820b485.png)

#### Output from manual plotting
![grafik](https://user-images.githubusercontent.com/21331242/110164918-5c5ba680-7df2-11eb-8b91-5f55cfffcdeb.png)

### Results
- There is an **issue with my Background fit CIs**. Im not sure what the issue is, but this is not the main problem. 
- The distance distribution is identical
- The fit to the trace (`Vfit`) is identical in both plots.
- There is an **offset in the experimental data** (`Vexp`)

## Output matching

This issue targets the last point. To me it seems that `fit.plot()` is correct (due to better fit), but manual output/plotting isn't. In this way it is not possible to plot/use the data in manually plotted figures.
The `fit.scale` property was only used for scaling of the `Bfit`. For plotting it was not necessary.

### Error origin
Found the error for the datapoint 'offset'. In the manual plotting, one has to replace the plotting (line 51 in MWE) command with
```python
axs[0].plot(t,Vexp/np.max(Vexp)*fit.scale,'.',color='grey',alpha=0.5)
```
to bypass the issue.

Reason for this is, that the scale factor is not equal to `np.max(Vexp)`. This is only the prescale (line 496)
```python
        prescales = [max(V) for V in Vexp]
```
The scale factor is then later (line 506) adjusted:
```python
        scales = [prescales[i]*np.trapz(Pfit,r) for i in range(nSignals)]
```
To obtain `Vfit`, `scales` is used for normalization (line 510). This then gives a discrepancy to the `Vexp` which is taken from outside/identical to input. In the internal plotting `Vexp` is first scaled by `prescales` (line line 497)and then lateron readjusted by `scales` to fit the `Vfit` in the plot (line 578). 
To me an easy solution is not obvious. But I'm not into your code. I do not understand, why two scaling factors are needed. In my opinion the output of `fit.V` and `fit.B`, etc. should be comparable to the `Vexp` input in the function.

_Originally posted by @laenan8466 in https://github.com/JeschkeLab/DeerLab/issues/74#issuecomment-791638127_

## Background CI scaling
Reason seems to be that for internal plotting a `scaledBfit_uq` is used and for output it is not.
The output is identical with the internal plotting when setting the `scalebckg=True` flag in `fitsignal`.
But then the plot has to be adjusted:
```python
axs[0].plot(t,(fit.B),'--',color='tab:orange')
```
and
```python
axs[0].fill_between(t,(scaledBfit95[:,0]), (scaledBfit95[:,1]),facecolor='tab:orange',linestyle='None',alpha=0.2)
axs[0].fill_between(t,(scaledBfit50[:,0]), (scaledBfit50[:,1]),facecolor='tab:orange',linestyle='None',alpha=0.4)
```

# Proposed solutions
Please comment on these propose solutions, as then one can work on a commit to fix them one by one.

## Output matching
I would like a behaviour for all things as follows (= means comparable with other stuff):
```
Input * Scalefactor = Scaled output
Input * 1 = UNscaled output
```
In this way for the user there are only two types of dataset (of course there could be needed more scales for internal processing):
- Scaled to 1
- Unscaled

From my point of view it seems that the scaled output should be scaled the internal scale (and **not** simply to max(Vexp). But I'm no expert here.

## Scaled and unscaled output
Was discussed in https://github.com/JeschkeLab/DeerLab/pull/78#issuecomment-791755000 and I have no further input.

## Background CI
This could be solved similar to *Output matching*. To me it is unclear if the background should start at 1 (or the scale factor) or fitted to the modulation depth. This could be discussed here. At the moment with
- **`scaledbckg = False`** *(default)* it is scaled to 1 and *not* corrected to modulation depth
- **`scaledbckg = True`** it is scaled to fit.scale and corrected to modulation depth

To me this seems unintuitive as scalebckg alters two settings at once. I have no opinion in this case how to improve.

**Comment by laenan8466** [comment]
S.th. different that came to my mind:
Why not include all inputs (so Vexp, t and r) in the `FitResult` class? Then everything would come together in a scaled/not scaled way.

Idea somehow extends on #31: One would have a result and a data input object. Maybe both could be combined.

**Comment by luisfabib** [comment]
>  Output doesn't match input. Scale factor output (fit.scale) does not scale the input (Vexp) correctly to match the fit.V ouput.
> As described in #74 (comment)

The recent changes in #78 now fixes this. All (time-domain) components of the signal are scaled as `Vexp`. So now the components can be plotted directly on top of `Vexp` without the need of scaling either of them.

> User should have the choice between a scaled and not scaled output.
> This builds up from #78 (comment). First attempt was made in fix/fitsignal_plot.

While we have contemplated this, we have decided that the operation is too basic to include it as a built-in functionality. To switch between scaling to `V0` or scaling to `1`, the components must just be divided by the scale in `FitResult.scale`. 

> Background CIs cannot be scaled from the given output in fit.
> This was mentioned in #74 (comment) as well.

This is also fixed in #78. There was a new poorly named output with the option `scaledBckg` which returns the unmodulated component of the signal (i.e. (1-lambda)*B in 4-pulse DEER). It had nothing to do with the overall scale `V0` of the signal. We have renamed this to an output `Vunmod` to avoid this confusion. 

Overall the scaling has been adjusted for all outputs and their uncertainties in `fitsignal` to be in accordance with `Vexp`.

**Comment by luisfabib** [comment]
> S.th. different that came to my mind:
> Why not include all inputs (so Vexp, t and r) in the `FitResult` class? Then everything would come together in a scaled/not scaled way.
> 
> Idea somehow extends on #31: One would have a result and a data input object. Maybe both could be co

This is still an open idea, but will most definitely not be included in the upcoming release. We are also contemplating similar designs for other parts of DeerLab. So this discussion will be open again in the short-term future for sure.
